File: Laba_8_c++(list)/List.cpp
#include <iostream>
#include"List.h"

using namespace std;



List::List()
    {
        Size = 0;
        head = nullptr;
    };
 void List::PushBack(int data)
    {
        if (head == nullptr)
        {
            head = new Node(data);
        }
        else
        {
            Node* current = this->head;
            while (current->pNext != nullptr)
            {
                current = current->pNext;
            }
            current->pNext = new Node(data);
        }
        Size++;
    }
    int List::GetSize()
    {
        return Size;
    }
    int& List:: operator[](const int index)
    {
        int counter = 0;
        Node* current = this->head;
        while (current != nullptr)
        {
            if (counter == index)
            {
                return current->data;
            }
            current = current->pNext;
            counter++;
        }
    }
    void List::PopFront()
    {
        Node* temp = head;
        head = head->pNext;
        delete temp;
        Size--;
    }
    void List:: PushFront(int data)
    {
        head = new Node(data, head);
        Size++;
    }
    void List:: Insert(int data, int index)
    {
        if (index == 0)
        {
            PushFront(data);
        }
        else
        {
            Node* previous = this->head;
            for (int i = 0; i < index - 1; i++)
            {
                previous = previous->pNext;
            }
            Node* NewNode = new Node(data, previous->pNext);
            previous->pNext = NewNode;
            Size++;
        }
    }
    void List:: RemoveAt(int index)
    {
        if (index == 0)
        {
            PopFront();
        }
        else
        {
            Node* previous = this->head;
            for (int i = 0; i < index - 1; i++)
            {
                previous = previous->pNext;
            }
            Node* RemoveEl = previous->pNext;
            previous->pNext = RemoveEl->pNext;
            delete RemoveEl;
            Size--;
        }
    }
    void List:: InsertAfterFirst(int data)
    {
        Insert(data, 1);
    }
    List:: ~List()
    {
        while (Size)
        {
            PopFront();
        }
    };

    List::Node::Node(int data , Node* pNext )
    {
        this->pNext = pNext;
        this->data = data;

    };
  

File: Laba_8cshrp(list)/List.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections;

namespace Lists
{
    public class List : IEnumerable
    {

        public class Node
        {
            public Node Next { get; set; }
            public int data { get; set; }
            public Node(int data, Node Next)
            {
                this.Next = Next;
                this.data = data;

            }
        };
        public int Size { get; private set; }
        public Node head { get; private set; }

        public List()
        {
            Size = 0;
            head = null;
        }
        public void PushBack(int data)
        {
            if (head == null)
            {
                head = new Node(data, null);
            }
            else
            {
                Node current = this.head;
                while (current.Next != null)
                {
                    current = current.Next;
                }
                current.Next = new Node(data, null);
            }
            Size++;
        }

        public void PopFront()
        {
            Node temp = head;
            head = head.Next;
            Size--;
        }
        public void PushFront(int data)
        {
            head = new Node(data, head);
            Size++;
        }
        public void Insert(int data, int index)
        {
            if (index == 0)
            {
                PushFront(data);
            }
            else
            {
                Node previous = this.head;
                for (int i = 0; i < index - 1; i++)
                {
                    previous = previous.Next;
                }
                Node NewNode = new Node(data, previous.Next);
                previous.Next = NewNode;
                Size++;
            }
        }
        public void RemoveAt(int index)
        {
            if (index == 0)
            {
                PopFront();
            }
            else
            {
                Node previous = this.head;
                for (int i = 0; i < index - 1; i++)
                {
                    previous = previous.Next;
                }
                Node RemoveEl = previous.Next;
                previous.Next = RemoveEl.Next;
                Size--;
            }
        }
        public void InsertAfterFirst(int data, int index)
        {
            Insert(data, 1);
        }
        IEnumerator IEnumerable.GetEnumerator()
        {
            Node current = head;
            while (current != null)
            {
                yield return current.data;
                current = current.Next;
            }
        }
        public int this[int index]
        {
            get
            {
                Node previous = head;
                for (int i = 0; i < index;i++)
                {
                    previous = previous.Next;
                }
                return previous.data;
            }
            set
            {
                Node previous = head;
                for (int i = 0; i < index; i++)
                {
                    previous = previous.Next;
                }
                previous.data = value;

            }
        }
        public int MultiplesOfTwo()
        {
            int MultiplesOfTwo = 0;
            for (int i = 0; i < this.Size; i++)
            {
                if (this[i] % 2 == 0 && i % 2 == 0)
                {
                    MultiplesOfTwo++;
                }
            }
            return MultiplesOfTwo;
        }
        public void RemoveAfterMin()
        {
            int MinIndex = 0;
            int Min = this[MinIndex];
            for (int i = 0; i < this.Size; i++)
            {
                if (this[i] < Min)
                {
                    Min = this[i];
                    MinIndex = i;
                }
            }
            int PreviousListSize = this.Size;
            for (int i = 0; i < PreviousListSize - MinIndex - 1; i++)
            {
                this.RemoveAt(MinIndex + 1);
            }
        }

    } 
}

